firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java

```diff
@@ -148,7 +157,18 @@ synchronized void storeHeartBeat(long millis, String userAgentString) {
     String dateString = getFormattedDate(millis);
     String lastDateString = firebaseSharedPreferences.getString(LAST_STORED_DATE, "");
     if (lastDateString.equals(dateString)) {
-      return;
+      String storedUserAgentString = getStoredUserAgentString(dateString);
+      if (storedUserAgentString == null) {
+        // Heartbeat already sent for today.
+        return;
+      }
+      if (storedUserAgentString.equals(userAgentString)) {
+        // UserAgent not updated.
+        return;
+      } else {
+        updateStoredUserAgent(userAgentString, dateString);
```

Comment:
how often can this happen? I'm thinking about the additional cost of reading (line 160), and writing(169), the user agent string.

Reply:
i think the updateStoredUserAgent doesn't happen as much. But the other case when the user agent remains the same might happen a lot which is why i think line 160 is necessary

Reply:
My understanding is this is called when `initializeApp()` is called and when the app is resumed from the background.
Does not seem to be too frequent.

Reply:
Reading `SharedPreferences` is `synchronized` so it can get expense, but maybe since we are already in a `synchronized` block is not too much worse? what impact do you think this change can have?